tcms/issuetracker/bitbucket.py

```diff
@@ -0,0 +1,140 @@
+                        'Content-type': 'application/json'}
+        self.auth = HTTPBasicAuth(api_username, api_password)
+
+    def create_issue(self, data):
+        url = '{0}{1}'.format(self.endpoint_url, '/issues')
+        return self._request("POST", url, headers=self.headers, auth=self.auth, json=data)
+
+    def get_issue(self, issue_id):
+        url = '{0}{1}{2}'.format(self.endpoint_url, '/issues/', issue_id)
+        return self._request("GET", url, headers=self.headers, auth=self.auth)
+
+    def update_issue(self, issue_id, data):
+        url = '{0}{1}{2}{3}'.format(self.endpoint_url, '/issues/', issue_id, '/changes')
+        return self._request("POST", url, headers=self.headers, auth=self.auth, json=data)
+
+    def add_comment(self, issue_id, comment):
+        url = '{0}{1}{2}{3}'.format(self.endpoint_url, '/issues/', issue_id, '/comments/')
+        return self._request("POST", url, headers=self.headers, auth=self.auth, json=comment)
+
+    def get_comments(self, issue_id):
+        url = '{0}{1}{2}{3}'.format(self.endpoint_url, '/issues/',
+                                    issue_id, '/comments?sort=-updated_on')
+        return self._request("GET", url, headers=self.headers, auth=self.auth)
+
+    @staticmethod
+    def _request(method, url, **kwargs):
+        return requests.request(method, url, **kwargs).json()
+
+    @staticmethod
+    def _construct_endpoint_url(api_version, url):
+        splitted_url = url.replace('https://', '').split('/')
+        base_url = "https://api.bitbucket.org"
+        workspace = splitted_url[1]
+        repository = splitted_url[2]
+        endpoint_url = "{0}/{1}/{2}/{3}/{4}".format(base_url,
+                                                    api_version,
+                                                    'repositories',
+                                                    workspace,
+                                                    repository)
+        return endpoint_url
+
+
+class BitBucketThread(IntegrationThread):
+    """
+        Execute BitBucket code in a thread!
+
+        Executed from the IssueTracker interface methods.
+    """
+
+    def post_comment(self):
+        comment_body = {"content": {"raw": self.text()}}
+        self.rpc.add_comment(self.bug_id, comment_body)
+
+
+class BitBucket(IssueTrackerType):
+    """
+        Support for BitBucket. Requires:
+
+        :base_url: - URL to a BitBucket project for which we're going to report issues
+        e.g. https://bitbucket.org/{workspace}/{repository}
+        :api_username: BitBucket Username
+        :api_password: BitBucket App Password
+
+        .. note::
+
+            You can leave the ``api_url`` field blank because
+            the integration code doesn't use them!
+    """
+    it_class = BitBucketThread
+
+    def _rpc_connection(self):
+        # NOTE: BitBucket 'app password' is required as api_password
+        # NOTE: BitBucket 'username' (not e-mail address) is required as api_username
```

Comment:
hmmm, is this the same username used to login into their website ? In other words you can't use the token by itself ?

Reply:
Yes its the same username which you creat the "app password". And its required to be given otherwise api call gets 401.

Check the "Using an app password" section here;
https://support.atlassian.com/bitbucket-cloud/docs/app-passwords/

Reply:
This is not very intuitive and the URL definitely needs to be in the doc string for the BitBucket class. I also see this in their documentation:

```
Using an app password

... skip ...
    username: your normal Bitbucket username
    password: the app password.
```

So yeah, we need to document it.

Reply:
I don't know how i can document it.

Reply:
You can add this in the doc-string for the BitBucket class. See around line 85 in that file, there is a note section added. You can add note/warning/info sections in a similar manner and they will be rendered with appropriate styling when published on ReadTheDocs.

Reply:
Ok, by the way i have noticed a problem in "Azure Boards" document [here](https://kiwitcms.readthedocs.io/en/latest/modules/tcms.issuetracker.azure_boards.html?highlight=AzureBoards#tcms.issuetracker.azure_boards.AzureBoards);

Check the line starting with "e.g. https://dev.azure.com/{organization}/{project} "

Which is caused by this [line](https://github.com/kiwitcms/Kiwi/blob/dbf5b2e8459fdb835d2e8e7a1e2788c60f884618/tcms/issuetracker/azure_boards.py#L74)

I will create a separate PR for it.

Reply:
Yep, items in doc-strings need to be aligned properly.